src/cdk-experimental/selection/selection.ts

```diff
@@ -0,0 +1,223 @@
+  private _deselectable: boolean = true;
+
+  /** The max number of selections that can be selected */
+  @Input('cdkSelectionMaxSelected')
+  get maxSelections(): number { return this._maxSelections; }
+  set maxSelections(val) {
+    if (val !== undefined && val !== null) {
+      this._maxSelections = coerceNumberProperty(val);
+    }
+  }
+  private _maxSelections: number;
+
+  /** Whether the control is disabled or not. */
+  @Input('cdkSelectionDisabled')
+  get disabled(): boolean { return this._disabled; }
+  set disabled(val) { this._disabled = coerceBooleanProperty(val); }
+  private _disabled: boolean;
+
+  /** Event fired when a selection/deselection occurs. */
+  @Output() selectionChange = new EventEmitter<SelectionChange<T>>();
+
+  /** The model backing of the component. */
+  _selectionModel: SelectionModel<T>;
+
+  /** Set of registered toggle components. */
+  _toggleDirectives: any[] = [];
+
+  /** Previously selected index used for range selection. */
+  _previousSelectedIndex = -1;
+
+  constructor(private _elementRef: ElementRef) {}
+
+  ngOnInit() {
+    this. _selectionModel = new SelectionModel<T>(
+      this.mode === 'multiple',
+      this.selectedItems,
+      true,
+      this.trackBy
+    );
+
+    if (this._selectionModel.change) {
+      this._selectionModel.change.subscribe(e => {
+        this.selectedItems = this._selectionModel.selected;
+        this.selectionChange.emit(e);
+      });
+    }
+  }
+
+  /** Register the toggle directive. */
+  registerToggle(toggle) {
+    this._toggleDirectives.push(toggle);
+  }
+
+  /** Deregister the toggle directivet. */
+  deregisterToggle(toggle) {
+    const toggleIndex = this._toggleDirectives.indexOf(toggle);
+    if (toggleIndex > -1) {
+      this._toggleDirectives.splice(toggleIndex, 1);
+    }
+  }
+
+  /** Toggle the selection of a item. */
+  toggle(toggle) {
+    if (this.disabled) {
+      return;
+    }
+
+    if (Array.isArray(toggle.model)) {
+      if (this.mode === 'multiple' &&
+          (this.maxSelections === undefined ||
+          (toggle.model.length + this._selectionModel.selected.length) < this.maxSelections)) {
+        this._selectionModel.clear();
+        this._selectionModel.toggle(...toggle.model);
+      }
+    } else {
+
+      this._setSelections(toggle);
+    }
+  }
+
+  /** Set the user selection for the element. */
+  setTextSelection(type: string) {
+    const elm = this._elementRef.nativeElement;
+    elm.style.userSelect = type;
+    elm.style.webkitUserSelect = type;
+    elm.style.MozUserSelect = type;
+  }
+
+  /** Set the selection state given a toggle directive and its activated options. */
+  private _setSelections(toggle) {
+    const ctrls = this._getSortedSet();
+    const index = this._getToggleIndex(ctrls, toggle);
+    const { modifier, model } = toggle;
+    const selections = this._selectionModel.selected;
+
+    if (this.mode === 'multiple' && modifier === 'shift') {
```

Comment:
Could the body of this method be broken up into two other methods: `_setMultipleSelections` and `_setsingleSelections`? Right now it feels like it has a lot going on.